adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -121,6 +125,102 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+   * @return a new SortedGenomicRDDMixIn with the RDD partitioned evenly
+   */
+  private def evenlyRepartition(partitions: Int)(implicit tTag: ClassTag[T]): GenomicRDD[T, U] = {
+    require(sorted, "Cannot evenly repartition an unsorted RDD")
+    // the average number of records on each node will help us evenly repartition
+    val average: Double = rdd.count.toDouble / partitions
+    // we already have a sorted rdd, so let's just move the data
+    // but we want to move it in blocks to maintain order
+    // zipWithIndex seems to be the cheapest way to guarantee this
+    val finalPartitionedRDD = rdd
+      .mapPartitionsWithIndex((idx, iter) => {
+        // here we get the destination partition number
+        // this will calculate so that each partition has a nearly
+        // equal amount of data on it
+        extractReferenceRegionsForPartition(iter, partitionMap.get(idx))
+      }, preservesPartitioning = true).zipWithIndex
+      .mapPartitionsWithIndex((idx, iter) => {
+        // divide the global index by the average to get the destination
+        // partition number
+        iter.map(_.swap).map(f => ((f._2._1, (f._1 / average).toInt), f._2._2))
+      })
+      .repartitionAndSortWithinPartitions(new ReferenceRegionRangePartitioner(partitions))
+
+    val finalPartitionMap = finalPartitionedRDD.mapPartitions(iter =>
+      getRegionBoundsFromPartition(iter.map(f => (f._1._1, f._2)))).collect
+
+    replaceRdd(finalPartitionedRDD.values, Some(finalPartitionMap))
+  }
+
+  /**
+   * Infers the correct reference region for each T given a the bounds of the partition (partitionMap)
+   * and returns the iterator with each T paired with the ReferenceRegion that it represents in that
+   * sorted location. This is important because a T can map to multiple ReferenceRegions.
+   *
+   * @param iter The iterator (that partition's data) of T objects
+   * @param partitionMap the bounds of the partition where _._1 is the lower bound and _._2 is the
+   *                     upper bound
+   * @return An iterator of all T's keyed with the specific ReferenceRegion that the T represents in
+   *         the global sort
+   */
+  private def extractReferenceRegionsForPartition(
+    iter: Iterator[T], partitionMap: Option[(ReferenceRegion, ReferenceRegion)]): Iterator[((ReferenceRegion, T))] = {
+
+    if (iter.isEmpty || partitionMap.isEmpty) {
+      Iterator()
+    } else {
+      // we put the records that map to multiple locations in this ListBuffer
+      // IMPORTANT: these are records that map to multiple locations on this
+      // partition. There should be very few of these
+      // we use this because we have to keep track of which ReferenceRegions
+      // have been used.
+      val duplicates = new ListBuffer[(Seq[ReferenceRegion], T)]()
+      // filter out ReferenceRegions that are not on this partition
+      iter.map(f => (getReferenceRegions(f).filter(g =>
```

Comment:
I'd like a small organizational cleanup here. Can you wrap the map function for mapping over the iterator in {}'s and then separate the filter from the map? I think this would make the code easier to follow.